reps/2022-09-19-ray-on-spark.md

```diff
@@ -0,0 +1,258 @@
+
+Using Spark's barrier mode, we intend to represent each Ray worker node as a long-running Spark
+task, where the Ray worker node has the same set of resources as the Spark task. Multiple Spark
+tasks can run on a single Spark worker node, which means that multiple Ray worker nodes can run
+on a single Spark worker node. Finally, we intend to run the Ray head node on the Spark driver
+node with a fixed resource allocation.
+
+To clarify further, the following example demonstrates how a Ray cluster with 16 total worker CPU
+cores can be launched on a Spark cluster, assuming that each Ray node requires 4 CPU cores and
+10GB of memory:
+
+1. On the Spark driver node, launch the Ray head node as follows:
+
+```
+ray start --head --num-cpus=0
+```
+
+The Ray head node will operate solely as a  manager node, without worker node functionality.
+
+2. Create a Spark barrier mode job, which executes all tasks in the spark job concurrently. Each
+task is allocated a fixed number of CPUs and a fixed amount of memory. In this example, we will
+allocate 4 CPU cores to each Spark task and at least 10GB of memory (by ensuring that each Spark
+worker node has at least 10GB of memory reserved for every set of 4 CPU cores, as computed by
+``SPARK_WORKER_MEMORY`` / ``TASK_SLOTS_PER_WORKER`` under the assumption that the
+``spark.task.cpus`` configuration is set to ``4``).
+
+3. In each constituent Spark task, launch one Ray worker node in blocking mode and allocate to it
+the full set of resources available to the Spark task (4 CPUs, 10GB of memory). Keep the Spark task
+running until the Ray cluster destroyed. The command used to start each Ray worker node is as
+follows:
+
+```
+ray start —head --num-cpus=X --num-gpus=Y --memory=Z --address={head_ip}:{port}
+```
+
+4. After the Ray cluster is launched, the user's Ray application(s) can be submitted to the Ray
+cluster via the Ray head node address / port.
+
+5. To shut down the Ray cluster, cancel the Spark barrier mode job and terminate all Ray node
+services that are running on the Spark driver and Spark workers.
+
+Finally, by adjusting the resource allocation for Ray nodes and Spark tasks, this approach enables
+users to run multiple Ray clusters in isolation of each other on a single Spark cluster, if desired.
+
+### Key questions
+
+#### We use spark barrier mode job to launch a ray cluster, shall we make the spark barrier to be a background job and keep running, until user explicitly terminate it ?
+
+Yes. So every user only need to start ray cluster once and then user can run ray applications
+on the ray cluster with little overhead.
+
+
+#### Launch Ray head node on spark driver node or spark task side ?
+In spark driver side.
+Because we can set Ray head node with `num-cpus` = 0, so that no ray task will be scheduled
+to ray head node, so that we can launch ray head node in driver side, it won't consume
+too much computing / memory resources of driver node.
```

Comment:
Based on the discussion, while the ray head node won't run any application code, it still runs critical system processes so we should make sure it runs in a production machine instead of like user's laptop.

Reply:
Excellent point. I'll update the REP and ensure that we perform validation on an instance with minimum available specs (similar to the change that we just pushed to the PR that performs a worker-level available resources check)